{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "d1875dd6",
   "metadata": {},
   "source": [
    "# üß¨ Discovering Emerging Topics in Drug Discovery Research using BERTopic"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "b8c61d45",
   "metadata": {},
   "source": [
    "## 1Ô∏è‚É£Import dependencies"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 34,
   "id": "0332f931",
   "metadata": {},
   "outputs": [],
   "source": [
    "import os\n",
    "import re, spacy\n",
    "import pandas as pd\n",
    "from Bio import Entrez, Medline\n",
    "from bertopic import BERTopic\n",
    "from transformers import pipeline\n",
    "from umap import UMAP\n",
    "from hdbscan import HDBSCAN\n",
    "from sentence_transformers import SentenceTransformer\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "16501b19",
   "metadata": {},
   "source": [
    "## 2Ô∏è‚É£ Data Collection\n",
    "I used PubMed  to fetch abstracts on topics like 'drug discovery', 'AI drug design', and 'molecular docking'.\n",
    "Store as `data/raw_publications.csv`. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "5c5ee524",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Number of PMIDs found: 1000\n",
      "(976, 5)\n",
      "Saved to data/raw_publications.csv\n"
     ]
    }
   ],
   "source": [
    "Entrez.email = \"paulafredrick26@gmail.com\"  # already set\n",
    "# 1. Search PubMed\n",
    "handle = Entrez.esearch(\n",
    "    db=\"pubmed\",\n",
    "    term=query,\n",
    "    retmax=max_results,\n",
    "    sort=\"relevance\"\n",
    ")\n",
    "search_results = Entrez.read(handle)\n",
    "handle.close()\n",
    "\n",
    "pmids = search_results[\"IdList\"]\n",
    "print(\"Number of PMIDs found:\", len(pmids))\n",
    "\n",
    "# 2. Fetch details\n",
    "handle = Entrez.efetch(\n",
    "    db=\"pubmed\",\n",
    "    id=\",\".join(pmids),\n",
    "    rettype=\"medline\",\n",
    "    retmode=\"text\"\n",
    ")\n",
    "records = list(Medline.parse(handle))\n",
    "handle.close()\n",
    "\n",
    "# 3. Build DataFrame\n",
    "rows = []\n",
    "for rec in records:\n",
    "    abstract = rec.get(\"AB\", \"\")\n",
    "    if not abstract:\n",
    "        continue\n",
    "    rows.append({\n",
    "        \"pmid\": rec.get(\"PMID\", \"\"),\n",
    "        \"title\": rec.get(\"TI\", \"\"),\n",
    "        \"abstract\": abstract,\n",
    "        \"journal\": rec.get(\"JT\", \"\"),\n",
    "        \"year\": rec.get(\"DP\", \"\")[:4]\n",
    "    })\n",
    "\n",
    "df = pd.DataFrame(rows)\n",
    "print(df.shape)\n",
    "df.head()\n",
    "    \n",
    "# 4. Save CSV\n",
    "os.makedirs(\"data\", exist_ok=True)\n",
    "df.to_csv(\"data/raw_publications.csv\", index=False)\n",
    "print(\"Saved to data/raw_publications.csv\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "29ef61c8",
   "metadata": {},
   "source": [
    "### Start from Here!!"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "0e787778",
   "metadata": {},
   "source": [
    "## 3Ô∏è‚É£ Data Cleaning & Preprocessing"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "id": "bced69d2",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>pmid</th>\n",
       "      <th>title</th>\n",
       "      <th>abstract</th>\n",
       "      <th>journal</th>\n",
       "      <th>year</th>\n",
       "      <th>clean_text</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>31487867</td>\n",
       "      <td>Molecular Docking: Shifting Paradigms in Drug ...</td>\n",
       "      <td>Molecular docking is an established in silico ...</td>\n",
       "      <td>International journal of molecular sciences</td>\n",
       "      <td>2019</td>\n",
       "      <td>molecular docking establish silico structureba...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>26205061</td>\n",
       "      <td>Molecular docking and structure-based drug des...</td>\n",
       "      <td>Pharmaceutical research has successfully incor...</td>\n",
       "      <td>Molecules (Basel, Switzerland)</td>\n",
       "      <td>2015</td>\n",
       "      <td>pharmaceutical research successfully incorpora...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>38594926</td>\n",
       "      <td>The Art and Science of Molecular Docking.</td>\n",
       "      <td>Molecular docking has become an essential part...</td>\n",
       "      <td>Annual review of biochemistry</td>\n",
       "      <td>2024</td>\n",
       "      <td>molecular docking essential structural biologi...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>34560276</td>\n",
       "      <td>Machine-learning methods for ligand-protein mo...</td>\n",
       "      <td>Artificial intelligence (AI) is often presente...</td>\n",
       "      <td>Drug discovery today</td>\n",
       "      <td>2022</td>\n",
       "      <td>artificial intelligence ai present new industr...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>34147204</td>\n",
       "      <td>Use of molecular docking computational tools i...</td>\n",
       "      <td>Molecular docking has become an important comp...</td>\n",
       "      <td>Progress in medicinal chemistry</td>\n",
       "      <td>2021</td>\n",
       "      <td>molecular docking important component drug dis...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "       pmid                                              title  \\\n",
       "0  31487867  Molecular Docking: Shifting Paradigms in Drug ...   \n",
       "1  26205061  Molecular docking and structure-based drug des...   \n",
       "2  38594926          The Art and Science of Molecular Docking.   \n",
       "3  34560276  Machine-learning methods for ligand-protein mo...   \n",
       "4  34147204  Use of molecular docking computational tools i...   \n",
       "\n",
       "                                            abstract  \\\n",
       "0  Molecular docking is an established in silico ...   \n",
       "1  Pharmaceutical research has successfully incor...   \n",
       "2  Molecular docking has become an essential part...   \n",
       "3  Artificial intelligence (AI) is often presente...   \n",
       "4  Molecular docking has become an important comp...   \n",
       "\n",
       "                                       journal  year  \\\n",
       "0  International journal of molecular sciences  2019   \n",
       "1               Molecules (Basel, Switzerland)  2015   \n",
       "2                Annual review of biochemistry  2024   \n",
       "3                         Drug discovery today  2022   \n",
       "4              Progress in medicinal chemistry  2021   \n",
       "\n",
       "                                          clean_text  \n",
       "0  molecular docking establish silico structureba...  \n",
       "1  pharmaceutical research successfully incorpora...  \n",
       "2  molecular docking essential structural biologi...  \n",
       "3  artificial intelligence ai present new industr...  \n",
       "4  molecular docking important component drug dis...  "
      ]
     },
     "execution_count": 32,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "nlp = spacy.load(\"en_core_web_sm\")\n",
    "\n",
    "def clean_text(text):\n",
    "    text = str(text)\n",
    "    # collapse whitespace\n",
    "    text = re.sub(r\"\\s+\", \" \", text)\n",
    "    # remove punctuation, make lowercase\n",
    "    text = re.sub(r\"[^\\w\\s]\", \"\", text.lower())\n",
    "    doc = nlp(text)\n",
    "    tokens = [\n",
    "        token.lemma_\n",
    "        for token in doc\n",
    "        if not token.is_stop and token.is_alpha\n",
    "    ]\n",
    "    return \" \".join(tokens)\n",
    "\n",
    "df = pd.read_csv(\"data/raw_publications.csv\")\n",
    "df[\"clean_text\"] = df[\"abstract\"].apply(clean_text)\n",
    "df.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e2b9d03f",
   "metadata": {},
   "source": [
    "## 4Ô∏è‚É£ Embedding Generation"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 35,
   "id": "9f4a4f4f",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "application/vnd.jupyter.widget-view+json": {
       "model_id": "466abed122e64c3cbe3b094e160b137f",
       "version_major": 2,
       "version_minor": 0
      },
      "text/plain": [
       "Batches:   0%|          | 0/31 [00:00<?, ?it/s]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "\n",
    "model = SentenceTransformer('all-MiniLM-L6-v2')\n",
    "\n",
    "embeddings = model.encode(\n",
    "    df['clean_text'], \n",
    "    show_progress_bar=True, \n",
    "    convert_to_numpy=True\n",
    ")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e4c4688e",
   "metadata": {},
   "source": [
    "## 5Ô∏è‚É£ Hyperparameter Tuning"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 36,
   "id": "6220edb0",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Experiment\n",
    "umap_params = [\n",
    "    {\"n_neighbors\": 5,  \"n_components\": 5},\n",
    "    {\"n_neighbors\": 10, \"n_components\": 5},\n",
    "    {\"n_neighbors\": 15, \"n_components\": 5},\n",
    "    {\"n_neighbors\": 5,  \"n_components\": 10},\n",
    "    {\"n_neighbors\": 10, \"n_components\": 10},\n",
    "    {\"n_neighbors\": 15, \"n_components\": 10},\n",
    "]\n",
    "\n",
    "hdbscan_params = [\n",
    "    {\"min_cluster_size\": 8},\n",
    "    {\"min_cluster_size\": 12},\n",
    "]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 37,
   "id": "646ada3a",
   "metadata": {},
   "outputs": [],
   "source": [
    "results = []\n",
    "\n",
    "for u in umap_params:\n",
    "    for h in hdbscan_params:\n",
    "\n",
    "        umap_model = UMAP(\n",
    "            n_neighbors=u[\"n_neighbors\"],\n",
    "            n_components=u[\"n_components\"],\n",
    "            min_dist=0.0,\n",
    "            metric=\"cosine\",\n",
    "            random_state=42\n",
    "        )\n",
    "\n",
    "        hdbscan_model = HDBSCAN(\n",
    "            min_cluster_size=h[\"min_cluster_size\"],\n",
    "            metric=\"euclidean\",\n",
    "            cluster_selection_method=\"eom\"\n",
    "        )\n",
    "\n",
    "        topic_model = BERTopic(\n",
    "            umap_model=umap_model,\n",
    "            hdbscan_model=hdbscan_model,\n",
    "            verbose=False\n",
    "        )\n",
    "\n",
    "        topics, _ = topic_model.fit_transform(df[\"clean_text\"], embeddings)\n",
    "\n",
    "        info = topic_model.get_topic_info()\n",
    "\n",
    "        n_topics = len(info[info.Topic != -1])\n",
    "        noise_docs = info[info.Topic == -1][\"Count\"].values[0] if -1 in info.Topic.values else 0\n",
    "\n",
    "        results.append({\n",
    "            \"n_neighbors\": u[\"n_neighbors\"],\n",
    "            \"n_components\": u[\"n_components\"],\n",
    "            \"min_cluster_size\": h[\"min_cluster_size\"],\n",
    "            \"n_topics\": n_topics,\n",
    "            \"noise_docs\": noise_docs\n",
    "        })\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 38,
   "id": "fef1525e",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>n_neighbors</th>\n",
       "      <th>n_components</th>\n",
       "      <th>min_cluster_size</th>\n",
       "      <th>n_topics</th>\n",
       "      <th>noise_docs</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>10</td>\n",
       "      <td>10</td>\n",
       "      <td>8</td>\n",
       "      <td>23</td>\n",
       "      <td>287</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>10</td>\n",
       "      <td>5</td>\n",
       "      <td>8</td>\n",
       "      <td>22</td>\n",
       "      <td>345</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>15</td>\n",
       "      <td>5</td>\n",
       "      <td>8</td>\n",
       "      <td>21</td>\n",
       "      <td>362</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>5</td>\n",
       "      <td>5</td>\n",
       "      <td>8</td>\n",
       "      <td>21</td>\n",
       "      <td>222</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>5</td>\n",
       "      <td>10</td>\n",
       "      <td>8</td>\n",
       "      <td>19</td>\n",
       "      <td>238</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>15</td>\n",
       "      <td>10</td>\n",
       "      <td>8</td>\n",
       "      <td>17</td>\n",
       "      <td>279</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>15</td>\n",
       "      <td>10</td>\n",
       "      <td>12</td>\n",
       "      <td>10</td>\n",
       "      <td>249</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>10</td>\n",
       "      <td>5</td>\n",
       "      <td>12</td>\n",
       "      <td>10</td>\n",
       "      <td>236</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>15</td>\n",
       "      <td>5</td>\n",
       "      <td>12</td>\n",
       "      <td>9</td>\n",
       "      <td>263</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>5</td>\n",
       "      <td>10</td>\n",
       "      <td>12</td>\n",
       "      <td>6</td>\n",
       "      <td>64</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>5</td>\n",
       "      <td>5</td>\n",
       "      <td>12</td>\n",
       "      <td>3</td>\n",
       "      <td>9</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>10</td>\n",
       "      <td>10</td>\n",
       "      <td>12</td>\n",
       "      <td>3</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "    n_neighbors  n_components  min_cluster_size  n_topics  noise_docs\n",
       "8            10            10                 8        23         287\n",
       "2            10             5                 8        22         345\n",
       "4            15             5                 8        21         362\n",
       "0             5             5                 8        21         222\n",
       "6             5            10                 8        19         238\n",
       "10           15            10                 8        17         279\n",
       "11           15            10                12        10         249\n",
       "3            10             5                12        10         236\n",
       "5            15             5                12         9         263\n",
       "7             5            10                12         6          64\n",
       "1             5             5                12         3           9\n",
       "9            10            10                12         3           0"
      ]
     },
     "execution_count": 38,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "results_df = pd.DataFrame(results)\n",
    "results_df.sort_values(\"n_topics\", ascending=False)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "05bc1c9f",
   "metadata": {},
   "source": [
    "## 6Ô∏è‚É£ Topic Modelling with BERTopic"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 39,
   "id": "fcdaf99e",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Topic</th>\n",
       "      <th>Count</th>\n",
       "      <th>Name</th>\n",
       "      <th>Representation</th>\n",
       "      <th>Representative_Docs</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>-1</td>\n",
       "      <td>219</td>\n",
       "      <td>-1_compound_study_inhibitor_drug</td>\n",
       "      <td>[compound, study, inhibitor, drug, molecular, ...</td>\n",
       "      <td>[inhibition janus kinase member jak family tyr...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>0</td>\n",
       "      <td>379</td>\n",
       "      <td>0_drug_docking_discovery_method</td>\n",
       "      <td>[drug, docking, discovery, method, molecular, ...</td>\n",
       "      <td>[conventional drug discovery approach expensiv...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>1</td>\n",
       "      <td>105</td>\n",
       "      <td>1_virus_drug_protease_antiviral</td>\n",
       "      <td>[virus, drug, protease, antiviral, viral, comp...</td>\n",
       "      <td>[emergence new variant raise concern effective...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>2</td>\n",
       "      <td>53</td>\n",
       "      <td>2_kinase_inhibitor_compound_cell</td>\n",
       "      <td>[kinase, inhibitor, compound, cell, cancer, mo...</td>\n",
       "      <td>[kinase key target approximately breast cancer...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>3</td>\n",
       "      <td>44</td>\n",
       "      <td>3_inhibitor_cancer_bind_compound</td>\n",
       "      <td>[inhibitor, cancer, bind, compound, molecular,...</td>\n",
       "      <td>[background delta play key role bcell signal t...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>4</td>\n",
       "      <td>43</td>\n",
       "      <td>4_ad_disease_compound_inhibitor</td>\n",
       "      <td>[ad, disease, compound, inhibitor, alzheimer, ...</td>\n",
       "      <td>[discovery novel multifunctional inhibitor tar...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>5</td>\n",
       "      <td>30</td>\n",
       "      <td>5_alphaglucosidase_compound_alphaamylase_study</td>\n",
       "      <td>[alphaglucosidase, compound, alphaamylase, stu...</td>\n",
       "      <td>[diabete mellitus dm multifactorial lifethreat...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>6</td>\n",
       "      <td>24</td>\n",
       "      <td>6_gpcrs_ligand_receptor_gpcr</td>\n",
       "      <td>[gpcrs, ligand, receptor, gpcr, agonist, allos...</td>\n",
       "      <td>[gproteincouple receptor gpcrs tractable drug ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>7</td>\n",
       "      <td>23</td>\n",
       "      <td>7_bacterial_antibiotic_inhibitor_aureus</td>\n",
       "      <td>[bacterial, antibiotic, inhibitor, aureus, gyr...</td>\n",
       "      <td>[betalactamase ampc general cause onset antibi...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>8</td>\n",
       "      <td>23</td>\n",
       "      <td>8_antimalarial_malaria_drug_parasite</td>\n",
       "      <td>[antimalarial, malaria, drug, parasite, new, a...</td>\n",
       "      <td>[identify novel lead compound drug discovery c...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>9</td>\n",
       "      <td>17</td>\n",
       "      <td>9_tuberculosis_tb_drug_mycobacterium</td>\n",
       "      <td>[tuberculosis, tb, drug, mycobacterium, antitb...</td>\n",
       "      <td>[tuberculosis tb infectious disease cause numb...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>10</td>\n",
       "      <td>16</td>\n",
       "      <td>10_cancer_drug_anticancer_computational</td>\n",
       "      <td>[cancer, drug, anticancer, computational, disc...</td>\n",
       "      <td>[conventional drug design embrace gene drug di...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "    Topic  Count                                            Name  \\\n",
       "0      -1    219                -1_compound_study_inhibitor_drug   \n",
       "1       0    379                 0_drug_docking_discovery_method   \n",
       "2       1    105                 1_virus_drug_protease_antiviral   \n",
       "3       2     53                2_kinase_inhibitor_compound_cell   \n",
       "4       3     44                3_inhibitor_cancer_bind_compound   \n",
       "5       4     43                 4_ad_disease_compound_inhibitor   \n",
       "6       5     30  5_alphaglucosidase_compound_alphaamylase_study   \n",
       "7       6     24                    6_gpcrs_ligand_receptor_gpcr   \n",
       "8       7     23         7_bacterial_antibiotic_inhibitor_aureus   \n",
       "9       8     23            8_antimalarial_malaria_drug_parasite   \n",
       "10      9     17            9_tuberculosis_tb_drug_mycobacterium   \n",
       "11     10     16         10_cancer_drug_anticancer_computational   \n",
       "\n",
       "                                       Representation  \\\n",
       "0   [compound, study, inhibitor, drug, molecular, ...   \n",
       "1   [drug, docking, discovery, method, molecular, ...   \n",
       "2   [virus, drug, protease, antiviral, viral, comp...   \n",
       "3   [kinase, inhibitor, compound, cell, cancer, mo...   \n",
       "4   [inhibitor, cancer, bind, compound, molecular,...   \n",
       "5   [ad, disease, compound, inhibitor, alzheimer, ...   \n",
       "6   [alphaglucosidase, compound, alphaamylase, stu...   \n",
       "7   [gpcrs, ligand, receptor, gpcr, agonist, allos...   \n",
       "8   [bacterial, antibiotic, inhibitor, aureus, gyr...   \n",
       "9   [antimalarial, malaria, drug, parasite, new, a...   \n",
       "10  [tuberculosis, tb, drug, mycobacterium, antitb...   \n",
       "11  [cancer, drug, anticancer, computational, disc...   \n",
       "\n",
       "                                  Representative_Docs  \n",
       "0   [inhibition janus kinase member jak family tyr...  \n",
       "1   [conventional drug discovery approach expensiv...  \n",
       "2   [emergence new variant raise concern effective...  \n",
       "3   [kinase key target approximately breast cancer...  \n",
       "4   [background delta play key role bcell signal t...  \n",
       "5   [discovery novel multifunctional inhibitor tar...  \n",
       "6   [diabete mellitus dm multifactorial lifethreat...  \n",
       "7   [gproteincouple receptor gpcrs tractable drug ...  \n",
       "8   [betalactamase ampc general cause onset antibi...  \n",
       "9   [identify novel lead compound drug discovery c...  \n",
       "10  [tuberculosis tb infectious disease cause numb...  \n",
       "11  [conventional drug design embrace gene drug di...  "
      ]
     },
     "execution_count": 39,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "umap_model = UMAP(n_neighbors=10, n_components=5, min_dist=0.0, metric='cosine')\n",
    "hdbscan_model = HDBSCAN(min_cluster_size=12, metric='euclidean', cluster_selection_method='eom')\n",
    "\n",
    "topic_model = BERTopic(umap_model=umap_model, hdbscan_model=hdbscan_model)\n",
    "topics, probs = topic_model.fit_transform(df['clean_text'], embeddings)\n",
    "topic_model.get_topic_info()\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "8b05eab6",
   "metadata": {},
   "source": [
    "##### Topic evolution over time"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 40,
   "id": "206f8eba",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Topic</th>\n",
       "      <th>Words</th>\n",
       "      <th>Frequency</th>\n",
       "      <th>Timestamp</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>-1</td>\n",
       "      <td>inhibitor, study, covalent, compound, design</td>\n",
       "      <td>13</td>\n",
       "      <td>2015</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>0</td>\n",
       "      <td>docking, drug, discovery, screening, molecular</td>\n",
       "      <td>33</td>\n",
       "      <td>2015</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>1</td>\n",
       "      <td>virus, dengue, hbv, pr, protease</td>\n",
       "      <td>3</td>\n",
       "      <td>2015</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>2</td>\n",
       "      <td>biphenyl, amide, inhibitor, indicate, value</td>\n",
       "      <td>1</td>\n",
       "      <td>2015</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>3</td>\n",
       "      <td>hadometdc, inhibitor, polyamine, nec, model</td>\n",
       "      <td>3</td>\n",
       "      <td>2015</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>114</th>\n",
       "      <td>3</td>\n",
       "      <td>kcal, caffeic, phytoconstituent, ehretia, cancer</td>\n",
       "      <td>2</td>\n",
       "      <td>2025</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>115</th>\n",
       "      <td>4</td>\n",
       "      <td>fisher, target, identification, aid, bioassays</td>\n",
       "      <td>1</td>\n",
       "      <td>2025</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>116</th>\n",
       "      <td>5</td>\n",
       "      <td>kaempferol, alphaamylase, alphaglucosidase, an...</td>\n",
       "      <td>2</td>\n",
       "      <td>2025</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>117</th>\n",
       "      <td>7</td>\n",
       "      <td>ftsz, pneumoniae, inhibitor, leprosy, murg</td>\n",
       "      <td>3</td>\n",
       "      <td>2025</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>118</th>\n",
       "      <td>9</td>\n",
       "      <td>tuberculosis, tb, drug, katg, mycobacterium</td>\n",
       "      <td>3</td>\n",
       "      <td>2025</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>119 rows √ó 4 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "     Topic                                              Words  Frequency  \\\n",
       "0       -1       inhibitor, study, covalent, compound, design         13   \n",
       "1        0     docking, drug, discovery, screening, molecular         33   \n",
       "2        1                   virus, dengue, hbv, pr, protease          3   \n",
       "3        2        biphenyl, amide, inhibitor, indicate, value          1   \n",
       "4        3        hadometdc, inhibitor, polyamine, nec, model          3   \n",
       "..     ...                                                ...        ...   \n",
       "114      3   kcal, caffeic, phytoconstituent, ehretia, cancer          2   \n",
       "115      4     fisher, target, identification, aid, bioassays          1   \n",
       "116      5  kaempferol, alphaamylase, alphaglucosidase, an...          2   \n",
       "117      7         ftsz, pneumoniae, inhibitor, leprosy, murg          3   \n",
       "118      9        tuberculosis, tb, drug, katg, mycobacterium          3   \n",
       "\n",
       "     Timestamp  \n",
       "0         2015  \n",
       "1         2015  \n",
       "2         2015  \n",
       "3         2015  \n",
       "4         2015  \n",
       "..         ...  \n",
       "114       2025  \n",
       "115       2025  \n",
       "116       2025  \n",
       "117       2025  \n",
       "118       2025  \n",
       "\n",
       "[119 rows x 4 columns]"
      ]
     },
     "execution_count": 40,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "topics_over_time = topic_model.topics_over_time(\n",
    "    df[\"clean_text\"],\n",
    "    df[\"year\"]\n",
    ")\n",
    "topics_over_time "
   ]
  },
  {
   "cell_type": "markdown",
   "id": "125c30b4",
   "metadata": {},
   "source": [
    "## 7Ô∏è‚É£ Zero-Shot Topic Labeling for Enhanced Interpretability\n",
    "To improve interpretability of the unsupervised BERTopic results, I applied a zero-shot classification model (BART-MNLI) to automatically assign semantic labels to the representative documents of each topic."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 41,
   "id": "9dd37963",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Device set to use cpu\n"
     ]
    }
   ],
   "source": [
    "classifier = pipeline(\"zero-shot-classification\", \n",
    "                      model=\"facebook/bart-large-mnli\")\n",
    "\n",
    "candidate_labels = [\n",
    "    \"AI drug design\",\n",
    "    \"molecular docking\",\n",
    "    \"protein folding\",\n",
    "    \"cancer therapeutics\",\n",
    "    \"genomics\",\n",
    "    \"drug target discovery\",\n",
    "    \"machine learning methods\"\n",
    "]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "id": "05a14d11",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{-1: 'molecular docking', 0: 'molecular docking', 1: 'molecular docking'}"
      ]
     },
     "execution_count": 22,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#Label one topic at a time\n",
    "topic_representatives = topic_model.get_representative_docs()\n",
    "\n",
    "zero_shot_labels = {}\n",
    "\n",
    "for topic, docs in topic_representatives.items():\n",
    "    text = docs[0]  # representative document\n",
    "    result = classifier(text, candidate_labels)\n",
    "    zero_shot_labels[topic] = result[\"labels\"][0]  # best label\n",
    "zero_shot_labels"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e7afa596",
   "metadata": {},
   "source": [
    "## 8 Evaluation & Visualization"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 42,
   "id": "92d8149b",
   "metadata": {},
   "outputs": [
    {
     "ename": "KeyError",
     "evalue": "'publication_date'",
     "output_type": "error",
     "traceback": [
      "\u001b[1;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[1;31mKeyError\u001b[0m                                  Traceback (most recent call last)",
      "File \u001b[1;32me:\\Condadistribution\\envs\\ml\\lib\\site-packages\\pandas\\core\\indexes\\base.py:3812\u001b[0m, in \u001b[0;36mIndex.get_loc\u001b[1;34m(self, key)\u001b[0m\n\u001b[0;32m   3811\u001b[0m \u001b[38;5;28;01mtry\u001b[39;00m:\n\u001b[1;32m-> 3812\u001b[0m     \u001b[38;5;28;01mreturn\u001b[39;00m \u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43m_engine\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mget_loc\u001b[49m\u001b[43m(\u001b[49m\u001b[43mcasted_key\u001b[49m\u001b[43m)\u001b[49m\n\u001b[0;32m   3813\u001b[0m \u001b[38;5;28;01mexcept\u001b[39;00m \u001b[38;5;167;01mKeyError\u001b[39;00m \u001b[38;5;28;01mas\u001b[39;00m err:\n",
      "File \u001b[1;32mpandas/_libs/index.pyx:167\u001b[0m, in \u001b[0;36mpandas._libs.index.IndexEngine.get_loc\u001b[1;34m()\u001b[0m\n",
      "File \u001b[1;32mpandas/_libs/index.pyx:196\u001b[0m, in \u001b[0;36mpandas._libs.index.IndexEngine.get_loc\u001b[1;34m()\u001b[0m\n",
      "File \u001b[1;32mpandas/_libs/hashtable_class_helper.pxi:7088\u001b[0m, in \u001b[0;36mpandas._libs.hashtable.PyObjectHashTable.get_item\u001b[1;34m()\u001b[0m\n",
      "File \u001b[1;32mpandas/_libs/hashtable_class_helper.pxi:7096\u001b[0m, in \u001b[0;36mpandas._libs.hashtable.PyObjectHashTable.get_item\u001b[1;34m()\u001b[0m\n",
      "\u001b[1;31mKeyError\u001b[0m: 'publication_date'",
      "\nThe above exception was the direct cause of the following exception:\n",
      "\u001b[1;31mKeyError\u001b[0m                                  Traceback (most recent call last)",
      "Cell \u001b[1;32mIn[42], line 4\u001b[0m\n\u001b[0;32m      2\u001b[0m topic_model\u001b[38;5;241m.\u001b[39mvisualize_barchart()\n\u001b[0;32m      3\u001b[0m topic_model\u001b[38;5;241m.\u001b[39mvisualize_hierarchy()\n\u001b[1;32m----> 4\u001b[0m topic_model\u001b[38;5;241m.\u001b[39mvisualize_topics_over_time(df[\u001b[38;5;124m'\u001b[39m\u001b[38;5;124mclean_text\u001b[39m\u001b[38;5;124m'\u001b[39m], \u001b[43mdf\u001b[49m\u001b[43m[\u001b[49m\u001b[38;5;124;43m'\u001b[39;49m\u001b[38;5;124;43mpublication_date\u001b[39;49m\u001b[38;5;124;43m'\u001b[39;49m\u001b[43m]\u001b[49m)\n",
      "File \u001b[1;32me:\\Condadistribution\\envs\\ml\\lib\\site-packages\\pandas\\core\\frame.py:4113\u001b[0m, in \u001b[0;36mDataFrame.__getitem__\u001b[1;34m(self, key)\u001b[0m\n\u001b[0;32m   4111\u001b[0m \u001b[38;5;28;01mif\u001b[39;00m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39mcolumns\u001b[38;5;241m.\u001b[39mnlevels \u001b[38;5;241m>\u001b[39m \u001b[38;5;241m1\u001b[39m:\n\u001b[0;32m   4112\u001b[0m     \u001b[38;5;28;01mreturn\u001b[39;00m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39m_getitem_multilevel(key)\n\u001b[1;32m-> 4113\u001b[0m indexer \u001b[38;5;241m=\u001b[39m \u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mcolumns\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mget_loc\u001b[49m\u001b[43m(\u001b[49m\u001b[43mkey\u001b[49m\u001b[43m)\u001b[49m\n\u001b[0;32m   4114\u001b[0m \u001b[38;5;28;01mif\u001b[39;00m is_integer(indexer):\n\u001b[0;32m   4115\u001b[0m     indexer \u001b[38;5;241m=\u001b[39m [indexer]\n",
      "File \u001b[1;32me:\\Condadistribution\\envs\\ml\\lib\\site-packages\\pandas\\core\\indexes\\base.py:3819\u001b[0m, in \u001b[0;36mIndex.get_loc\u001b[1;34m(self, key)\u001b[0m\n\u001b[0;32m   3814\u001b[0m     \u001b[38;5;28;01mif\u001b[39;00m \u001b[38;5;28misinstance\u001b[39m(casted_key, \u001b[38;5;28mslice\u001b[39m) \u001b[38;5;129;01mor\u001b[39;00m (\n\u001b[0;32m   3815\u001b[0m         \u001b[38;5;28misinstance\u001b[39m(casted_key, abc\u001b[38;5;241m.\u001b[39mIterable)\n\u001b[0;32m   3816\u001b[0m         \u001b[38;5;129;01mand\u001b[39;00m \u001b[38;5;28many\u001b[39m(\u001b[38;5;28misinstance\u001b[39m(x, \u001b[38;5;28mslice\u001b[39m) \u001b[38;5;28;01mfor\u001b[39;00m x \u001b[38;5;129;01min\u001b[39;00m casted_key)\n\u001b[0;32m   3817\u001b[0m     ):\n\u001b[0;32m   3818\u001b[0m         \u001b[38;5;28;01mraise\u001b[39;00m InvalidIndexError(key)\n\u001b[1;32m-> 3819\u001b[0m     \u001b[38;5;28;01mraise\u001b[39;00m \u001b[38;5;167;01mKeyError\u001b[39;00m(key) \u001b[38;5;28;01mfrom\u001b[39;00m\u001b[38;5;250m \u001b[39m\u001b[38;5;21;01merr\u001b[39;00m\n\u001b[0;32m   3820\u001b[0m \u001b[38;5;28;01mexcept\u001b[39;00m \u001b[38;5;167;01mTypeError\u001b[39;00m:\n\u001b[0;32m   3821\u001b[0m     \u001b[38;5;66;03m# If we have a listlike key, _check_indexing_error will raise\u001b[39;00m\n\u001b[0;32m   3822\u001b[0m     \u001b[38;5;66;03m#  InvalidIndexError. Otherwise we fall through and re-raise\u001b[39;00m\n\u001b[0;32m   3823\u001b[0m     \u001b[38;5;66;03m#  the TypeError.\u001b[39;00m\n\u001b[0;32m   3824\u001b[0m     \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39m_check_indexing_error(key)\n",
      "\u001b[1;31mKeyError\u001b[0m: 'publication_date'"
     ]
    }
   ],
   "source": [
    "\n",
    "topic_model.visualize_topics()\n",
    "topic_model.visualize_barchart()\n",
    "topic_model.visualize_hierarchy()\n",
    "topic_model.visualize_topics_over_time(df['clean_text'], df['publication_date'])\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "053e801a",
   "metadata": {},
   "source": [
    "## 8Ô∏è‚É£ Results & Insights\n",
    "- List and interpret top topics.\n",
    "- Identify emerging or declining research themes.\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "088ebce8",
   "metadata": {},
   "source": [
    "## 9Ô∏è‚É£ Conclusion\n",
    "- Summarize key findings.\n",
    "- Highlight business and scientific relevance.\n",
    "- Suggest future work (e.g., comparing journals, regions, or institutions).\n"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "ml",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.19"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
